rest-api.md

```diff
@@ -0,0 +1,767 @@
+
+Query:
+
+- `used` - filter for used/unused keys. Possible values: true/false;
+- `operatorIndex` - filter for keys of operator with index `operatorIndex`;
+
+Response:
+
+Response depends on `module type`
+
+```typescript
+class Response {
+  data: {
+    keys: Key[];
+    module: Module;
+  };
+  meta: {
+    elBlockSnapshot: ELBlockSnapshot;
+  };
+}
+
+interface HttpException {
+  statusCode: number;
+  message: string;
+}
+
+class TooEarlyResponse implements HttpException {
+  statusCode: number = 425;
+  message: string = 'Too early response';
+}
+
+class NotFoundException implements HttpException {
+  statusCode: number = 404;
+}
+```
+
+Example:
+
+```
+curl -X 'GET' \
+  'http://localhost:3000/v1/modules/1/keys?used=true&operatorIndex=1' \
+  -H 'accept: application/json'
+```
+
+:::warning
+If API returns 425 code, it means database is not ready for work
+:::
+
+**POST** `/v1/modules/{module_id}/keys/find`
+
+`module_id` - staking router module contact address or id;
+
+Returns all keys found in db.
+
+Request body:
+
+```typescript
+interface RequestBody {
+  // public keys in lowercase
+  pubkeys: string[];
+}
+```
+
+Response:
+
+```typescript
+class Response {
+  data: {
+    keys: Key[];
+    module: Module;
+  };
+  meta: {
+    elBlockSnapshot: ElBlockSnapshot;
+  };
+}
+
+interface HttpException {
+  statusCode: number;
+  message: string;
+}
+
+class TooEarlyResponse implements HttpException {
+  statusCode: number = 425;
+  message: string = 'Too early response';
+}
+
+class NotFoundException implements HttpException {
+  statusCode: number = 404;
+}
+```
+
+Example:
+
+```
+curl -X 'POST' \
+  'http://localhost:3000/v1/modules/1/keys/find' \
+  -H 'Content-Type: application/json' \
+  -d '{
+  "pubkeys": [
+    "pubkey"
+  ]
+}'
+```
+
+:::warning
+If API returns 425 code, it means database is not ready for work
+:::
+
+### /validators
+
+**GET** `/v1/modules/{module_id}/validators/validator-exits-to-prepare/{operator_id}`
+
+`module_id` - staking router module contact address or id;
+
+This endpoint will return N of oldest lido validators for earliest epoch when voluntary exit can be processed for specific node operator and specific `StakingRouter` module. Node operator will use this validators list for preparing pre-sign exit messages. API will find `used` keys of node operator and find for these public keys N oldest validators in `Validator` table. We consider active validators (`active_ongoing` status) or validators in `pending_initialized`, `pending_queued` statuses. Module tables state fetched from `EL` should be newer than `Validator` table state fetched from `CL`. Otherwise API will return error on request.
+
+Query:
+
+Only one filter is available. If both parameters are provided, `percent` has a high priority.
+
+- `percent` - Percent of validators to exit. Default value is 10.
+- `max_amount` - Number of validators to exit. If validators number less than amount, return all validators.
```

Comment:
usually this param named `limit` - this is just part of pagination

Reply:
Yeah. It works like pagination, so could use pagination instead of this param.
Just for future - less params - better